### ISSUE TITLE
/V1/products/{sku}/media does not assign roles correctly

### ISSUE THREAD

**Issue by giacmir** [magento/magento2#21736]
When uploading media gallery images via API the last inserted one gets all roles (small_image, base, thumbnail) even if not required to and another image already has them assigned.

### Preconditions (*)

1. Magento 2.3.0
2. php 7.1

### Steps to reproduce (*)

1. create a new product without images
2. POST to /V1/products/{sku}/media a payload that includes all the three image types "base", "small_image", "thumbnail"
3.Verify in backend that the new image has all three roles
4. POST another image to /V1/products/{sku}/media, this time the "types" node should be an empty array

### Expected result (*)

1. The last inserted image should have no roles
2.The first image should have maintained its assigned roles

### Actual result (*)

1. last image has all three roles even if no one was specified.

**Comment by magento-engcom-team** [comment]
Hi @giacmir. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [x] Summary of the issue
- [x] Information on your environment
- [x] Steps to reproduce
- [x] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@giacmir do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [x] yes
- [ ] no

**Comment by giacmir** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @giacmir. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @giacmir, here is your [Magento instance](https://i-21736-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-21736-2-3-develop.instances.magento-community.engineering/admin](https://i-21736-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

-  _Next steps are available in case you are a member of **Community Maintainers**._

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
Hi @giacmir thank you for you report, this issue has already fixed in 2.3-develop branch.
![DeepinScreenshot_select-area_20190314110417](https://user-images.githubusercontent.com/40756439/54344093-ee3bd900-4648-11e9-8f03-5578cfeda085.png)

**Comment by giacmir** [comment]
Hi @engcom-backlog-nazar I actually tried that on the provided 2.3-develop instance yesterday and reproduced it. Was it fixed today?

**Comment by ghost** [comment]
@giacmir  can you share with your payloads ?

**Comment by ghost** [comment]
@giacmir 
![DeepinScreenshot_select-area_20190314111928](https://user-images.githubusercontent.com/40756439/54345136-1a585980-464b-11e9-906a-3bef811eeffb.png)
![DeepinScreenshot_select-area_20190314111940](https://user-images.githubusercontent.com/40756439/54345137-1a585980-464b-11e9-8556-a9615800a7f4.png)

**Comment by giacmir** [comment]
@engcom-backlog-nazar sure thing:

payload 1
```json
{
  "entry": {
    "media_type": "image",
    "label": "First",
    "position": "0",
    "disabled": 1,
    "types": [
      "thumbnail",
      "image",
      "small_image"
    ],
    "file": "first_file.jpg",
    "content": {
      "base64_encoded_data": "<base64 encoded image goes here>",
      "type": "image\/jpeg",
      "name": "first_file.jpg"
    },
    "extension_attributes": []
  }
}
```

payload 2
```json
{
  "entry": {
    "media_type": "image",
    "label": "Second",
    "position": "1",
    "disabled": 0,
    "types": [],
    "file": "second_file.jpg",
    "content": {
      "base64_encoded_data": "<base64 data goes here>",
      "type": "image\/jpeg",
      "name": "second_file.jpg"
    },
    "extension_attributes": []
  }
}
```

Thank you

**Comment by ghost** [comment]
@giacmir this only appears if you set all 3 types, but with 2 types works correct. and only for for first 2 images.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-backlog-nazar 
 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-98681` were created

**Issue Available**: @engcom-backlog-nazar, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by Al3ks73** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @Al3ks73. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @Al3ks73, here is your [Magento instance](https://i-21736-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-21736-2-3-develop.instances.magento-community.engineering/admin](https://i-21736-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by magento-engcom-team** [comment]
Hi @shikhamis11. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if you want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by irajneeshgupta** [comment]
I am working on this at #dmcdindia1

**Comment by irajneeshgupta** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @irajneeshgupta. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @irajneeshgupta, here is your [Magento instance](https://i-21736-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-21736-2-3-develop.instances.magento-community.engineering/admin](https://i-21736-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by irajneeshgupta** [comment]
This issue has already been resolved in 2.3.1

**Comment by engcom-Golf** [comment]
Hi @giacmir, I could not reproduce this issue on 2.4-develop branch
It was probably fixed with changes in PR #18952
Will close this ticket.

**Comment by magento-engcom-team** [comment]
Unfortunately, we are archiving this ticket now as it did not get much attention from both Magento Community and Core developers for an extended period. This is done in an effort to create a quality, community-driven backlog which will allow us to allocate the required attention more easily.

Please feel free to comment or reopen according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines) 
the ticket if you are still facing this issue on the latest `2.x-develop` branch. Thank you for collaboration.
